src/cargo/ops/cargo_install.rs

```diff
@@ -182,16 +197,48 @@ fn install_one(
             }
         }
         src.update()?;
-        select_pkg(src, krate, vers, config, false, &mut |path| {
+        select_pkg(&mut src, krate, None, config, false, &mut |path| {
             path.read_packages()
         })?
     } else {
+        let mut source = map.load(source_id, &HashSet::new())?;
+        let vers = if let Some(vers) = vers {
+            Some(parse_semver_req(vers)?)
+        } else if source.source_id().is_registry() {
+            Some(VersionReq::any())
+        } else {
+            None
+        };
+        if krate.is_some() && !no_track {
+            if let Some(vers) = vers.clone() {
+                if vers.is_exact() {
+                    if let Ok(pkg) =
+                        select_pkg(&mut source, krate, Some(vers), config, false, &mut |_| {
+                            bail!("(Ignored)")
+                        })
+                    {
+                        let (_ws, rustc, target) =
+                            make_ws_rustc_target(&config, opts, &source_id, pkg.clone())?;
+                        if let Ok(installed) = is_installed(&pkg, &rustc, &target) {
```

Comment:
This can be simplified with:
```suggestion
                        if let Ok(true) = is_installed(&pkg, &rustc, &target) {
```